### PR TITLE
fix: avoid endless loop when handling symlinks (#109) by @q0rban

### LINKED ISSUE #108: Problem with symlinks inside node_modules
### What version of pkg are you using?

5.16.1

### What version of Node.js are you using?

18.x

### What operating system are you using?

macOS

### What CPU architecture are you using?

arm64

### What Node versions, OSs and CPU architectures are you building for?

any

### Describe the Bug

vercel/pkg#1555 is still an issue on this fork.

Original description of the bug:

> The library [node-rdkafka ](https://github.com/Blizzard/node-rdkafka/)has four symlinks in their code. If you install this package and try to package the code into a binary, the build freezes at the point where it reaches the symlinks. Nothing happens after this, you need to hit Ctrl+C.
> 
> ```
> [debug]  Adding asset : ....
> <path-to-project>/node_modules/node-rdkafka/build/deps/librdkafka++.so
> [debug] Content of <path-to-project>/node_modules/node-rdkafka/build/deps/librdkafka++.so.1 is added to queue. It was required from <path-to-project>/package.json
> ^C
> richards@richards:<path-to-project>$
> ```

### Expected Behavior

> The build should succeed with the symlinks resolved correctly to the files inside the binary. Most likely the issue is that the symlink files are simply copied into the binary, still referencing to the old files.
> 
> Example:
> /path-to-project-on-my-local-machine/node_modules/node-rdkafka/build/deps/librdkafka++.so points to path-to-project-on-my-local-machine/node_modules/node-rdkafka/build/deps/librdkafka++.so.1 on my local machine. Once built, the librdkafka++.so must not point to path-to-project-on-my-local-machine/node_modules/node-rdkafka/build/deps/librdkafka++.so.1 anymore, but to something like /snapshot/path-to-project/node_modules/node-rdkafka/build/deps/librdkafka++.so.1, so inside the virtual file system of the binary.

### To Reproduce

> See https://github.com/Mall0c/node-rdkafka-pkg-demo.

### PR DESCRIPTION
**Pull request by q0rban**
This resolves #108 for me. I tried to run tests locally to see if this breaks anything, but I guess the tests need to run in CI as they were failing with or without this change on my environment.

Fixes #108

### CODE REVIEW

**Comment by robertsLando** [comment]
@q0rban I'm not sure that just removing code is the correct way to handle this. You should handle the case of infinite loop like keeping track of paths to be sure you don't hit the same path twice to break the loop

**Comment by q0rban** [comment]
The interesting thing is that the code is already keeping track of the paths. I couldn't figure out a scenario where this code wouldn't result in an infinite loop, which is why I deleted it. But then I realize it is assuming that the paths are not nested at the same level. So I believe the code will work if `path/to/some/symlink/index.js` is symlinked to `path/to/real/index.js` (since they have different nesting levels), but I believe it will always loop if `path/to/symlink/index.js` is linked to `path/to/real/index.js` since they have the same nesting depth. 

Furthermore, I don't believe this function will properly handle the scenario where there are multiple symlinks involved. For example, let's say module `chicken` is symlinked wholesale into another module:

```
path/to/node_modules/some-module/node_modules/symchicken -> path/to/node_modules/some-module/node_modules/chicken
```

but then, what happens if `chicken` has some links in it as well?

```
chicken/lib/link.js -> chicken/deeply/nested/index.js
```

This would result in

```
path/to/node_modules/some-module/node_modules/symchicken/lib/index.js -> path/to/node_modules/some-module/node_modules/chicken/deeply/nested/index.js
```

Does that mean this logic needs to recurse? The mind reels.

**Comment by q0rban** [comment]
Incidentally, do you have any tips on running tests locally? I don't see any documentation on that in the repo.

**Comment by q0rban** [comment]
@robertsLando updated the approach and would appreciate your thoughts. The existing logic seems error prone, and thus difficult to ascertain the intent of it.

**Comment by robertsLando** [comment]
See https://github.com/yao-pkg/pkg/pull/111/files

File: lib/walker.ts [hunk 1]
```diff
@@ -317,14 +317,23 @@ function stepDetect(
   }
 }
 
-function findCommonJunctionPoint(file: string, realFile: string) {
+async function findCommonJunctionPoint(file: string, realFile: string) {
```
Comment (robertsLando) [inline lib/walker.ts:320]: add a jsdoc here to explain what this function does

File: lib/walker.ts [hunk 2]
```diff
@@ -399,10 +409,15 @@ class Walker {
...
-    file = a.file;
-    realFile = a.realFile;
+  async appendSymlink(file: string, realFile: string) {
+    try {
+      const a = await findCommonJunctionPoint(file, realFile);
+      file = a.file;
+      realFile = a.realFile;
+    } catch (error) {
+      log.error((error as Error).message);
+      return;
```
[the remark this answers is not in the document]
Comment (robertsLando) [reply lib/walker.ts:418]: Not sure here, maybe it would be better to throw? This could create inconsistent binaries

**Review by robertsLando — CHANGES_REQUESTED** [review]

**Comment by robertsLando** [comment]
Also please fix lint issues: `yarn fix`

File: lib/walker.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (q0rban) [reply lib/walker.ts]: I agree. I've removed the error handling here.

File: lib/walker.ts [hunk 3]
```diff
@@ -317,14 +317,37 @@ function stepDetect(
...
 
-  return { file, realFile };
+    // If the directory is /, break out of the loop and log an error.
+    if (
+      file === path.parse(file).root ||
+      realFile === path.parse(realFile).root
+    ) {
+      throw new Error(
+        'Reached root directory without finding a common junction point',
+      );
```
Comment (robertsLando) [inline lib/walker.ts:348]: may be worth adding a reference of the two original paths in the error

**Review by robertsLando — COMMENTED** [review]
Could you also add a test that covers this? 

`test-99-#108`

**Comment by q0rban** [comment]
> Could you also add a test that covers this?
> 
> `test-99-#108`

Whoops, apologies, I missed this. Will take a look at what this entails.

**Review by robertsLando — CHANGES_REQUESTED** [review]
See review above

**Comment by q0rban** [comment]
Okay, test added. I can confirm that the test just hangs on the main branch, and builds properly on this branch.

File: lib/walker.ts (hunk 3 above)
Comment (q0rban) [reply lib/walker.ts:348]: We do not have the original paths anymore at this point.

**Comment by q0rban** [comment]
> See review above

I know, I requested another review after pushing fixes based on that review. :)

**Review by robertsLando — APPROVED** [review]
Thanks for this! LGTM :)

**Comment by robertsLando** [comment]
Found improvements on how test works? 😆 Spent 2 days trying to improve them

**Pull request merged**